### PR TITLE
power: enhance device power management

### LINKED ISSUE #17838: state DEVICE_PM_LOW_POWER_STATE of Device Power Management
Currently Device Power Management only support state DEVICE_PM_ACTIVE_STATE and DEVICE_PM_SUSPEND_STATE. And DEVICE_PM_SUSPEND_STATE will be triggered when system will enter into deep sleep state. But for non deep sleep state, device power management will not be considered. State of DEVICE_PM_LOW_POWER_STATE requires less time than DEVICE_PM_SUSPEND_STATE, so I suggest to schedule state DEVICE_PM_LOW_POWER_STATE for devices when system will suspend but not deep sleep.

### PR DESCRIPTION
**Pull request by wentongwu**
For distributed method of device power management, when system going to suspend state, not directly changing the state of device driver, just notify device driver system power management
state, device driver will make their own decisions based on system power management state and their own knowledge accordingly.
And for central method, when system go to sleep state (not deep sleep), make devices go to DEVICE_PM_LOW_POWER_STATE state which needs less time than DEVICE_PM_SUSPEND_STATE to save more power in the sleep state based on the policy which can be defined by the end user accordingly.

Fixes: #17838.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

### CODE REVIEW

File: include/device.h [hunk 1]
```diff
@@ -450,6 +451,29 @@ static inline int device_get_power_state(struct device *device,
...
+ * in that system.
+ *
+ * @param device Pointer to device structure of the driver instance.
+ * @param device_power_state Device power state to be filled by the device.
+ * @param ticks The upcoming kernel idle time.
+ *
+ * @retval 0 If successful.
+ * @retval Errno Negative errno code if failure.
+ */
+static inline int device_get_scheduled_power_state(struct device *device,
```
Comment (pizi-nordic) [inline include/device.h]: I am not sure that power state selection should be done by device drivers. The driver does not have knowledge about system state so it will be hard to implement reasonable algorithm there.

File: subsys/power/power.c [hunk 2]
```diff
@@ -101,16 +101,17 @@ enum power_states _sys_suspend(s32_t ticks)
...
 #if CONFIG_DEVICE_POWER_MANAGEMENT
-		/* Suspend peripherals. */
-		if (sys_pm_suspend_devices()) {
-			LOG_ERR("System level device suspend failed!");
-			sys_pm_notify_power_state_exit(pm_state);
-			pm_state = SYS_POWER_STATE_ACTIVE;
-			return pm_state;
-		}
+	/* Suspend peripherals. */
+	if (sys_pm_suspend_devices(ticks)) {
```
Comment (pizi-nordic) [inline subsys/power/power.c]: This is the part we should really think about IMHO.
At the moment we are coupling device power state with the system state, but I am not sure if we should do that:
- The devices which are already not used should be suspended earlier to save the power.
- The devices might need time to suspend/resume, and this part blocks whole system until this is done.
- We should not resume devices which are not used as this will increase power consumption.

Comment (wentongwu) [reply subsys/power/power.c]: > This is the part we should really think about IMHO.
> At the moment we are coupling device power state with the system state, but I am not sure if we should do that:
> 
> * The devices which are already not used should be suspended earlier to save the power.
> * The devices might need time to suspend/resume, and this part blocks whole system until this is done.
> * We should not resume devices which are not used as this will increase power consumption.

yes, but that should be done by device driver, peripheral should decide suspend/resume by itself, that will be triggered by himself's logic(for example, there is no traffic for wifi chipset connected to AP/MCU, wifi chipset will go to suspend state). This PR is that system decide to suspend, sys_pm_suspend_devices is to notify peripherals, at least peripheral know system's state, if it want to work, wakeup cpu first. Further peripheral can enter low power state according system state or system suspend ticks. Yes, it may take some time and block the whole system, but we have to do that(linux do the notify also one by one), let end users to tuning system suspend ticks and the time of devices entering lower state. 
And currently we let devices enter suspend state only when system going to deep sleep, but what about system in sleep state(not deep sleep)? I think we should let peripherals know system state at least.
I'm not familiar with the design of Zephyr's power management, maybe there is something wrong with what I said. @pizi-nordic

File: include/device.h (hunk 1 above)
Comment (wentongwu) [reply include/device.h]: can we just pass the state system going to or suspend ticks to devices? @pizi-nordic

File: subsys/power/power.c (hunk 2 above)
Comment (wentongwu) [reply subsys/power/power.c]: > This is the part we should really think about IMHO.
> At the moment we are coupling device power state with the system state, but I am not sure if we should do that:
> 
>     * The devices which are already not used should be suspended earlier to save the power.
> 
>     * The devices might need time to suspend/resume, and this part blocks whole system until this is done.
> 
>     * We should not resume devices which are not used as this will increase power consumption.

Reviewed power management code more, I think my patch doesn't follow Zephyr power management infrastructure. But I still think we should consider device power management even when system going to sleep state(not deep sleep) in order to save more power if only enable central method device power management.  
```
Central method
==============

In this method device power management is mostly done inside
:code:`sys_suspend()` along with entering a CPU or SOC power state.

If a decision to enter deep sleep is made, the implementation would enter it
only after checking if the devices are not in the middle of a hardware
transaction that cannot be interrupted. This method can be used in
implementations where the applications and components using devices are not
expected to be power aware and do not implement device power management.
```
My basic idea is to let device enter DEVICE_PM_LOW_POWER_STATE state in that case, and resume when system also resume to active state, also will add basic policy function for this logic, end user can adjust it based on the SOC they use. And this logic will be under marco !CONFIG_DEVICE_IDLE_PM. 

Really appreciate for your comments. @pizi-nordic

File: subsys/power/power.c [hunk 3]
```diff
@@ -116,6 +117,17 @@ enum power_states _sys_suspend(s32_t ticks)
 		 * in deep sleep mode.
 		 */
 		_sys_pm_idle_exit_notification_disable();
+#if CONFIG_DEVICE_POWER_MANAGEMENT
+	} else if (sys_pm_policy_low_power_devices(pm_state)) {
+		/* low power peripherals. */
+		if (sys_pm_low_power_devices()) {
```
Comment (vanti) [inline subsys/power/power.c]: @wentongwu  My two cents... I started looking at power management recently, and based on what I have seen thus far, I tend to agree with the comments from @pizi-nordic. My general impression is that the centralized method isn't very useful by itself for the reasons @pizi-nordic stated. It is trying to couple device power states with system power states, which I don't think is the typical scenario.

For optimal power management, we want to keep devices in their low-power state for as long as possible, not just when the CPU goes to sleep/deep-sleep. While it is useful for a device to know about the CPU's power transitions (e.g. for saving/restoring state that might be impacted), it is rarely the determining factor in changing a device's power state.

I see the "centralized method" more as a way to notify a device of system power states, so that it gets a chance to save any context if necessary, rather than initiating an actual device state change. What do you think of replacing calls to sys_pm_suspend_devices() and sys_pm_low_power_devices() with 'sys_pm_notify_device(pm_state)', which would be more in tune with how I think it should work?

Other more general thoughts:

- If the device is busy doing something that requires the CPU being active, my understanding is that it should inform the power manager with a call to device_busy_set(). It'd be nicer if there is a form of this API that passes in the system power state the device cares about (e.g. device_busy_set(DISALLOW_SYS_POWER_STATE_DEEP_SLEEP_2)). The device might want a way to say "I am too busy to deal with the CPU entering deep sleep #2, but sleep #1 is ok"). The suspend logic should ensure that it doesn't perform a CPU power transition a device doesn't wish to occur. It is unclear to me whether the suspend logic is taking busy states into consideration today.

- One other thing that seems missing is power management at the power domain level. It'd be great to automatically turn off unused power domains on the CPU while it is in active state, and have a way for the devices declare their dependencies on these. In this manner, if none of the devices used in an application depends on a power domain, it can safely be shut off.

Comment (wentongwu) [reply subsys/power/power.c]: @vanti @pizi-nordic Thanks for the review and sorry for the late response. I mostly agree with your idea and also the power domain consideration is also in the plan. I will firstly add the notification mechanism to have devices know CPU's power state. The return value will be the indication whether devices is happy with that state or not, but doesn't consider device's suggestion currently. And I will keep the "centralized method" in case anyone need it, like the power button and in Android system. How do you think of that? Thanks

Comment (erwango) [reply subsys/power/power.c]: I didn't read the discussion in details, but I see interest in some use cases for "centralized method". This could be used, for some architectures lacking pm features (like LP timer), by application to manage a pm scenario. Application can decide to enter the low power state with `sys_suspend` after it has configured proper ways to wake up (like RTC alarm, GPIOs). So I'm good with keeping the "centralized method".

Comment (pizi-nordic) [reply subsys/power/power.c]: In my oppinion, the centralized control is something which should be implemented layer above. So instead of linking System power management to device power management we should have a mechanism which order both to reach the equivalent power state at the same time.

Comment (vanti) [reply subsys/power/power.c]: @wentongwu I am fine with keeping the "centralized method" if there are use-cases for it. My concern is that looking at the code it seems to be the only supported method today when CONFIG_DEVICE_POWER_MANAGEMENT=y is set, which gives the impression that device power state transitions are supposed to be initiated by system power state transitions. So yes having a notification mechanism and a way to turn off the "centralized method" would be helpful.

Comment (wentongwu) [reply subsys/power/power.c]: @pizi-nordic @vanti Could you please give some comments for the developing patch first? And it's about the notification mechanism, centralized method will be considered more. Thanks.

**Review by vanti — CHANGES_REQUESTED** [review]
@wentongwu Thanks for putting this together. I added a few comments.

File: subsys/power/power.c [hunk 4]
```diff
@@ -101,16 +101,16 @@ enum power_states _sys_suspend(s32_t ticks)
 	post_ops_done = 0;
 	sys_pm_notify_power_state_entry(pm_state);
 
-	if (deep_sleep) {
 #if CONFIG_DEVICE_POWER_MANAGEMENT
-		/* Suspend peripherals. */
-		if (sys_pm_suspend_devices()) {
```
[the remark this answers is not in the document]
Comment (wentongwu) [reply subsys/power/power.c]: @vanti I don't think current solution is a complete implementation of "centralized method", there is only suspending(power gate) devices, if some platform requires more fine-grained schedule(e.g. clock gate also needed), which is also what I'm trying to add initially, current code has no way helping that.

**Comment by wentongwu** [comment]
starting test - test_timer_duration_period

    Assertion failed at ZEPHYR_BASE/tests/kernel/timer/timer_api/src/main.c:61: duration_expire: k_uptime_delta(&tdata.timestamp) >= 100 is false

FAIL - test_timer_duration_period
=============

Not my PR cause this failure.

**Comment by wentongwu** [comment]
@vanti @pizi-nordic can you please take a look the patch? Thanks

File: subsys/power/policy/pm_policy.h [hunk 5]
```diff
@@ -18,11 +18,23 @@ extern "C" {
...
+extern int sys_pm_power_state_notify_devices(enum power_states state);
+
+#else
 /**
  * @brief Function to suspend the devices in PM device list
  */
 extern int sys_pm_suspend_devices(void);
 
+/**
+ * @brief Function to low power the devices in PM device list
```
[the remark this answers is not in the document]
Comment (wentongwu) [reply subsys/power/policy/pm_policy.h]: @vanti Thanks, I will change it. 
But how do you think the way handling the central method and distributed method?

File: subsys/power/power.c [hunk 6]
```diff
@@ -101,16 +101,39 @@ enum power_states _sys_suspend(s32_t ticks)
...
+	if (deep_sleep) {
 		/* Suspend peripherals. */
 		if (sys_pm_suspend_devices()) {
 			LOG_ERR("System level device suspend failed!");
+			sys_pm_resume_devices();
 			sys_pm_notify_power_state_exit(pm_state);
 			pm_state = SYS_POWER_STATE_ACTIVE;
 			return pm_state;
 		}
+	} else if (sys_pm_policy_low_power_devices(pm_state)) {
```
[the remark this answers is not in the document]
Comment (wentongwu) [reply subsys/power/power.c]: I just have end users control the policy easily, sys_pm_is_sleep_state is a example, maybe there is some other policies. 
I'm happy to make changes to look better.

**Comment by erwango** [comment]
@FRASTM

File: subsys/power/power.c [hunk 7]
```diff
@@ -101,16 +101,39 @@ enum power_states _sys_suspend(s32_t ticks)
...
 	sys_pm_notify_power_state_entry(pm_state);
 
-	if (deep_sleep) {
 #if CONFIG_DEVICE_POWER_MANAGEMENT
+#ifdef CONFIG_DEVICE_IDLE_PM
+	if (sys_pm_power_state_notify_devices(pm_state)) {
+		LOG_ERR("Notify devices system power state failed!");
+		sys_pm_notify_power_state_exit(pm_state);
+		pm_state = SYS_POWER_STATE_ACTIVE;
+		sys_pm_power_state_notify_devices(pm_state);
```
Comment (vanti) [inline subsys/power/power.c]: There could be a potential problem here. Say I have 7 devices. If sys_pm_power_state_notify_devices() failed when calling device_pm_notify() on the 3rd device, I don't think we need to notify the last 4 devices that we are going back to SYS_POWER_STATE_ACTIVE.

File: subsys/power/device.c [hunk 8]
```diff
@@ -41,10 +41,30 @@ static const char core_devices[NUM_CORE_DEVICES][MAX_DEV_NAME_LEN] = {
  * device power policies would be executed.
  */
 static int device_ordered_list[MAX_PM_DEVICES];
-static int device_retval[MAX_PM_DEVICES];
 static struct device *pm_device_list;
 static int device_count;
 
+#ifdef CONFIG_DEVICE_IDLE_PM
+int sys_pm_power_state_notify_devices(enum power_states state)
```
Comment (vanti) [inline subsys/power/device.c]: @wentongwu Just FYI: there is a good app note on the power manager for TI SimpleLink devices you may wish to read: http://software-dl.ti.com/simplelink/esd/simplelink_cc13x2_26x2_sdk/3.20.00.68/exports/docs/tidrivers/Power_Management.pdf. There are some concepts in there (e.g. notifications, constraints, dependencies, resources) that you may find interesting. The source code is BSD licensed, and the ideas are generally applicable for other platforms.

Comment (wentongwu) [reply subsys/power/device.c]: @vanti Thanks.

**Comment by wentongwu** [comment]
@pizi-nordic also appreciate for your inputs. Thanks

File: subsys/power/power.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (vanti) [reply subsys/power/power.c]: I'd have thought sys_pm_low_power_devices() is where you tell the device to do clock gating or enter some low-power state, not sys_pm_policy_low_power_devices()...? The latter is calling sys_pm_is_sleep_state() which only checks if the sleep state is valid.

[the remark this answers is not in the document]
Comment (wentongwu) [reply subsys/power/power.c]: yes, you are right, sorry, my mistake. should be sys_pm_low_power_devices to notify clock gating. 
for sys_pm_policy_low_power_devices, I just leave a interface to let end users decide which sleep state(1, 2, 3) can do clock gating.

File: subsys/power/power.c (hunk 7 above)
Comment (wentongwu) [reply subsys/power/power.c]: @vanti I will consider this more. could you please review other changes first? Thanks

**Review by erwango — CHANGES_REQUESTED** [review]
Please update zephyr documentation to reflect your addition

File: subsys/power/power.c (hunk 7 above)
Comment (wentongwu) [reply subsys/power/power.c]: @vanti as below code, before resume we will check the state of devices. Sorry for the late response. 
```
void sys_pm_resume_devices(void)
{
        int i;

        for (i = 0; i < device_count; i++) {
                if (!device_retval[i]) {
                        int idx = device_ordered_list[i];

                        device_set_power_state(&pm_device_list[idx],
                                        DEVICE_PM_ACTIVE_STATE, NULL, NULL);
                }
        }
}
```

**Comment by wentongwu** [comment]
> Please update zephyr documentation to reflect your addition

yes, I will update documentation if most people accept this patch. Could you please give a review for the code first? Thanks

**Review by pabigot — COMMENTED** [review]
I'm seeing several changes here:

* addition of a Kconfig distinction between centralized and distributed power management
* addition of API to support notifications of state changes for distributed-only
* addition of API related to clock gate devices for centralized-only

These could be three separate commits for clarity.  The third certainly seems worth being its own commit where it can be explained.

Style: there's no need for `extern` on the function declarations in the header.  However, that's pre-existing style and you've stayed consistent, so if you choose to remove them it should be in a separate commit, and if you don't then they're fine IMO.

I don't feel comfortable approving at this time because I'm still waiting on things like #19454 to support proper ordering for device-based power management.  But I don't see anything that concerns me from a technical perspective at this time.

File: subsys/power/device.c [hunk 9]
```diff
@@ -97,6 +138,7 @@ void sys_pm_resume_devices(void)
 		}
```
[the remark this answers is not in the document]
Comment (wentongwu) [reply subsys/power/device.c]: @vanti I don't think devices should reject the resume request from system IMHO.

File: subsys/power/power.c [hunk 10]
```diff
@@ -101,16 +101,39 @@ enum power_states _sys_suspend(s32_t ticks)
 	post_ops_done = 0;
 	sys_pm_notify_power_state_entry(pm_state);
 
-	if (deep_sleep) {
 #if CONFIG_DEVICE_POWER_MANAGEMENT
+#ifdef CONFIG_DEVICE_PM_DISTRIBUTED_METHOD
+	if (sys_pm_power_state_notify_devices(pm_state)) {
+		LOG_ERR("Notify devices system power state failed!");
+		sys_pm_notify_power_state_exit(pm_state);
```
[the remark this answers is not in the document]
Comment (wentongwu) [reply subsys/power/power.c]: Thanks

File: subsys/power/device.c [hunk 11]
```diff
@@ -66,6 +89,24 @@ int sys_pm_suspend_devices(void)
 	return 0;
 }
 
+int sys_pm_clock_gate_devices(void)
```
[the remark this answers is not in the document]
Comment (wentongwu) [reply subsys/power/device.c]: @vanti I prefer common function. Thanks

File: subsys/power/device.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (wentongwu) [reply subsys/power/device.c]: @vanti Now for CENTRAL METHOD, I think devices should follow system's request, shouldn't vote for the decisions made by system, that's the meaning of "central method", what do you think about this idea?

[the remark this answers is not in the document]
Comment (wentongwu) [reply subsys/power/device.c]: but for DISTRIBUTED_METHOD, devices can vote to the system state transition.

**Comment by wentongwu** [comment]
@pabigot thanks for the review and suggestion. could you please take a look updated patches? Thanks

**Review by nashif — APPROVED** [review]

**Review by pabigot — COMMENTED** [review]
tl;dr: Update the user documentation and confirm nothing depends on the changed behavior and I'll approve this.

OK, I think I get what this is doing.  I have a feeling that a full implementation of distributed device power management will require some changes, but we're probably at least a couple Zephyr releases away from getting that in place anyway.

Unless I'm mistaken: For existing behavior to be preserved any device that implements `device_pm_control` and currently could return a non-zero value to indicate that transition to `DEVICE_PM_SUSPEND_STATE` is rejected must be modified to instead invoke `sys_pm_suspend_lock_acquire()/release()` around any activity that inhibits that transition.

The APDS9660 sensor is the only device I've found that could be affected by this (@jfischer-phytec-iot), and in that situation it's probably wrong to block the suspension.  Be aware I have not done an exhaustive search; see summary below for what was checked.

If nothing blocks system suspend this way the PR seems to retain equivalent functionality to what it replaces while adding support for an intermediate system power state, so it's a "good thing".

There's already a request from @erwango to update the user documentation.  I do want to see that update before I will approve this.

In that documentation the concept of "gating" should be explained for those of us who are software-oriented.  As I reconstruct it (based on Wikipedia):

> When a clock is "gated" the driving clock signal is blocked from entering circuitry associated with some function.  This reduces the power consumption within that circuitry while retaining configuration and some state.
> When power is "gated" current is withdrawn from circuitry associated with some function.  This eliminates power consumption within that circuitry, causing loss of configuration and state.
> Control of clock gating saves energy but allows the function to be enabled and disabled quickly.
> Control of power gating saves more energy than clock-gating but increases the time required to enable or disable the function.

(If that text is accurate you're welcome to use it; it's original to me, not plagiarized from anywhere.  It may also be partially or completely wrong.)

From what I see "clock-gate the device" is being treated as equivalent to "put device in `DEVICE_PM_LOW_POWER_STATE`".  The terminology inconsistency is confusing.  Is clock gating the only technology that can do that?  If not, can we use another name, e.g. replace `<prefix>_clock_gate_devices` with `<prefix>_set_devices_to_low_power`?  Or even `<prefix>_low_power_devices`, treating "low-power" as a verb meaning "to set to low-power" (which English technically allows; you can verb anything in English).

I'm not going to insist on that rename because it's very difficult to do given the mix of adjectives and verbs currently used to name the power management states and functions.  But this is a pretty large inconsistency between the function name and its effect, given that "suspend" and "resume" match one-to-one.

For my own notes here's my reconstruction of what's going on.
* First commit is cleanup, no problem.
* Second commit removes the existing distributed (cooperative) device power management approach and creates a new central power management approach.  The difference is that with central control devices are told to adjust to the new device state unconditionally, while the previous approach allowed devices to influence the system state (i.e. override the final system power state) by asking them if it was ok to suspend.  The behavior of the existing "sys_pm_force_suspend" becomes the only supported behavior.
* Third commit adds API specific to the central PM approach, to decide based on the selected power state whether to "clock gate" all the devices.  Clock-gating propagates the low-power state down to the devices without allowing them to override it, just as suspend propagates the suspend state down to the devices.
* Fourth commit distinguishes centralized and distributed device power management, and adds a mechanism where state controlled by device power management can block entry to the lowest power state.  This replaces the "ask-for-approval" approach that was removed in the second commit.  It assumes that devices automatically go to at least low-power state when they are not active.

The following drivers have been identified as implementing `device_pm_control`.  The list may not be complete.
* drivers/gpio/gpio_dw: never rejects (deep)
* drivers/i2c/i2c_nrfx_twi: never rejects
* drivers/i2c/i2c_nrfx_twim: never rejects
* drivers/interrupt_controller/arcv2_irq_unit: never rejects (deep)
* drivers/interrupt_controller/ioapic_intr: never rejects (deep)
* drivers/interrupt_controller/loapic_intr: never rejects (deep)
* drivers/pwm/pwm_nrfx: never rejects
* drivers/sensor/apds9960/apds9960.c: **May reject, based on the result of an I2C transaction.**
* drivers/sensor/qdec_nrfx: never rejects
* drivers/serial/uart_nrfx_uart: never rejects
* drivers/serial/uart_nrfx_uarte: never rejects
* drivers/spi/spi_nrfx_spi: never rejects
* drivers/spi/spi_nrfx_spim: never rejects
* drivers/timer/sys_clock_init: never rejects
* samples/subsys/power/device_pm/src/dummy_driver: never rejects
* samples/subsys/power/device_pm/src/dummy_parent: never rejects

**Comment by wentongwu** [comment]
@pabigot @vanti  thanks for the review, I need more time to consider your comments, and currently I'm busying with tracing work. Thanks a lot.
